### PR TITLE
Add support for SET ROLE statement

### PR DESCRIPTION
**Pull request by ulisesojeda**
Add support for SET ROLE statement

Fixes https://github.com/trinodb/trino-python-client/issues/7

Follow-up of closed PR https://github.com/trinodb/trino-python-client/pull/196.

### CODE REVIEW

File: trino/client.py [hunk 1]
```diff
@@ -241,6 +241,7 @@ class TrinoRequest(object):
     :request_timeout: How long (in seconds) to wait for the server to send
                       data before giving up, as a float or a
                       ``(connect timeout, read timeout)`` tuple.
+    :param role: role for the current session. Some connectors do not support
```
Comment (mdesmet) [inline trino/client.py:244]: Please complete the sentence.

File: trino/client.py [hunk 2]
```diff
@@ -288,11 +289,13 @@ def __init__(
         request_timeout: Union[float, Tuple[float, float]] = constants.DEFAULT_REQUEST_TIMEOUT,
         handle_retry=exceptions.RetryWithExponentialBackoff(),
         verify: bool = True,
+        role: Optional[str] = None,
     ) -> None:
         self._client_session = client_session
         self._host = host
         self._port = port
         self._next_uri: Optional[str] = None
+        self._role = role
```
Comment (mdesmet) [inline trino/client.py:298]: role should be applied onto `ClientSession`, the set role will apply to the connection and must be thread-safe.

`ClientSession` is initialized in the `dbapi.Connection` and further modified based on the headers coming back from the server.

See how schema and catalog are modified as an example.

File: tests/integration/test_dbapi_integration.py [hunk 3]
```diff
@@ -32,6 +32,15 @@ def trino_connection(run_trino):
     )
 
 
+@pytest.fixture
+def trino_connection_jmx(run_trino):
```
Comment (mdesmet) [inline tests/integration/test_dbapi_integration.py:36]: Not sure if you need to add this new connection. See Trino docs: https://trino.io/docs/current/sql/set-role.html

File: tests/integration/test_dbapi_integration.py [hunk 4]
```diff
@@ -1034,3 +1043,27 @@ def test_use_catalog(run_trino):
     result = cur.fetchall()
     assert result[0][0] == 'tpch'
     assert result[0][1] == 'sf1'
+
+
+@pytest.mark.skipif(trino_version() == '351', reason="Newer Trino versions return the system role")
+def test_set_role_trino_higher_351(trino_connection_jmx):
+    cur0 = trino_connection_jmx.cursor()
```
Comment (mdesmet) [inline tests/integration/test_dbapi_integration.py:1050]: Nit: rename `cur0` to `cur`

File: tests/integration/test_dbapi_integration.py [hunk 5]
```diff
@@ -1034,3 +1043,27 @@ def test_use_catalog(run_trino):
...
+    assert cur0._request.role is None
+
+    cur0.execute("SET ROLE ALL")
+    cur0.fetchall()
+    assert cur0._request.role == "system=ALL"
+
+
+@pytest.mark.skipif(trino_version() != '351', reason="Trino 351 returns the role for the current catalog")
+def test_set_role_trino_351(trino_connection_jmx):
+    cur0 = trino_connection_jmx.cursor()
```
Comment (mdesmet) [inline tests/integration/test_dbapi_integration.py:1062]: Nit: rename `cur0` to `cur`

File: tests/unit/test_client.py [hunk 6]
```diff
@@ -68,7 +68,7 @@ def test_request_headers(mock_get_and_post):
     client_info_header = constants.HEADER_CLIENT_INFO
     client_info_value = "some_client_info"
 
-    req = TrinoRequest(
+    req_without_role = TrinoRequest(
```
Comment (mdesmet) [inline tests/unit/test_client.py:71]: Please split the tests per scenario. That makes it so much easier to troubleshoot if problems arise. Avoid numbering and try to make the name of the scenario as explicit as possible.

File: trino/client.py (hunk 1 above)
Comment (ulisesojeda) [reply trino/client.py]: :+1:

File: trino/client.py (hunk 2 above)
Comment (ulisesojeda) [reply trino/client.py]: Sure thing. Thanks for the example

File: tests/integration/test_dbapi_integration.py (hunk 3 above)
Comment (ulisesojeda) [reply tests/integration/test_dbapi_integration.py:36]: you're right. Not needed

File: tests/integration/test_dbapi_integration.py (hunk 4 above)
Comment (ulisesojeda) [reply tests/integration/test_dbapi_integration.py]: :+1:

**Comment by mdesmet** [comment]
Please note that you should not merge master into your PR, that way **existing** changes are re-introduced in your PR with a new commit id. Instead you should rebase your branch on master (pull latest changes from upstream repo on master and then `git rebase master` on your PR branch).

**Review by mdesmet — COMMENTED** [review]
Good work!

LGTM % comments, rebase and commit squash

**Review by mdesmet — APPROVED** [review]
LGTM. Please squash the commits

**Comment by ulisesojeda** [comment]
> LGTM. Please squash the commits

Ready. Thanks

File: tests/unit/test_client.py [hunk 7]
```diff
@@ -1002,3 +1002,195 @@ def __call__(self):
...
+    _, get_kwargs = get.call_args
+    assert_headers_with_role(
+        post_kwargs["headers"],
+        catalog,
+        schema,
+        source,
+        user,
+        accept_encoding_value,
+        client_info_value,
+        "hive=ALL"
```
Comment (ebyhr) [inline tests/unit/test_client.py:1081]: Please rewrite tests to focus on role change. For instance, encoding is unrelated to role tests.

Comment (ulisesojeda) [reply tests/unit/test_client.py]: Ready

File: tests/unit/test_client.py [hunk 8]
```diff
@@ -1002,3 +1002,95 @@ def __call__(self):
     with_retry(FailerUntil(2).__call__)()
     with pytest.raises(SomeException):
         with_retry(FailerUntil(3).__call__)()
+
+
+def assert_headers_with_role(headers, role):
+    catalog = "test_catalog"
+    schema = "test_schema"
+    user = "test_user"
+    source = "test_source"
```
Comment (ebyhr) [inline tests/unit/test_client.py:1011]: Remove `catalog`, `schema` and `source` as it unrelates to role tests.

File: tests/unit/test_client.py [hunk 9]
```diff
@@ -1002,3 +1002,95 @@ def __call__(self):
...
+def test_request_headers_role_hive_all(mock_get_and_post):
+    get, post = mock_get_and_post
+    req = TrinoRequest(
+        host="coordinator",
+        port=8080,
+        client_session=ClientSession(
+            user="test_user",
+            source="test_source",
+            catalog="test_catalog",
+            schema="test_schema",
```
Comment (ebyhr) [inline tests/unit/test_client.py:1031]: Remove `catalog`, `schema` and `source` as it unrelates to role tests. Please fix `test_request_headers_role_admin` & `test_request_headers_role_empty` as well.

File: tests/unit/test_client.py [hunk 10]
```diff
@@ -1002,3 +1002,95 @@ def __call__(self):
...
+            role="hive=ALL",
+        ),
+        http_scheme="http",
+        redirect_handler=None,
```
Comment (ebyhr) [inline tests/unit/test_client.py:1035]: Remove `http_scheme` and `redirect_handler` as it unrelates to role tests. Please fix `test_request_headers_role_admin` & `test_request_headers_role_empty` as well.

**Review by ebyhr — COMMENTED** [review]
Please squash commits into one.

File: tests/unit/test_client.py [hunk 11]
```diff
@@ -94,7 +94,7 @@ def assert_headers(headers):
         assert headers[constants.HEADER_SESSION] == ""
```
Comment (mdesmet) [inline tests/unit/test_client.py:94]: As the count of headers increased, please also assert the role is empty string.

```suggestion
        assert headers[constants.HEADER_SESSION] == ""
        assert headers[constants.HEADER_ROLE] == ""
```

File: tests/unit/test_client.py (hunk 10 above)
Comment (ulisesojeda) [reply tests/unit/test_client.py]: :+1:

File: tests/unit/test_client.py (hunk 11 above)
Comment (ulisesojeda) [reply tests/unit/test_client.py:94]: Done. Thanks

File: tests/unit/test_client.py (hunk 9 above)
Comment (ulisesojeda) [reply tests/unit/test_client.py]: :+1:

File: tests/unit/test_client.py (hunk 8 above)
Comment (ulisesojeda) [reply tests/unit/test_client.py]: :+1:

**Review by ebyhr — APPROVED** [review]

**Comment by ebyhr** [comment]
Merged, thanks!

**Comment by ulisesojeda** [comment]
> Merged, thanks!

Thanks!

**Pull request merged**